wave_utils/src/math.cpp

```diff
@@ -3,343 +3,314 @@
-  R << 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, cos(angle), sin(angle), 0.0f, 0.0f,
-    -sin(angle), cos(angle), 0.0f, 0.0f, 0.0f, 0.0f, 1.0f;
-
-  return 0;
-}
+int euler2rot(Vec3 euler, int euler_seq, Mat3 &R) {
+    double R11, R12, R13, R21, R22, R23, R31, R32, R33;
+    double phi, theta, psi;
+
+    phi = euler(0);
+    theta = euler(1);
+    psi = euler(2);
+
+    if (euler_seq == 321) {
+        // euler 3-2-1
+        R11 = cos(theta) * cos(psi);
+        R12 = sin(phi) * sin(theta) * cos(psi) - cos(phi) * sin(psi);
+        R13 = cos(phi) * sin(theta) * cos(psi) + sin(phi) * sin(psi);
+
+        R21 = cos(theta) * sin(psi);
+        R22 = sin(phi) * sin(theta) * sin(psi) + cos(phi) * cos(psi);
+        R23 = cos(phi) * sin(theta) * sin(psi) - sin(phi) * cos(psi);
+
+        R31 = -sin(theta);
+        R32 = sin(phi) * cos(theta);
+        R33 = cos(phi) * cos(theta);
+
+    } else if (euler_seq == 123) {
+        // euler 1-2-3
+        R11 = cos(theta) * cos(psi);
+        R12 = cos(theta) * sin(psi);
+        R13 = -sin(theta);
+
+        R21 = sin(phi) * sin(theta) * cos(psi) - cos(phi) * sin(psi);
+        R22 = sin(phi) * sin(theta) * sin(psi) + cos(phi) * cos(psi);
+        R23 = sin(phi) * cos(theta);
+
+        R31 = cos(phi) * sin(theta) * cos(psi) + sin(phi) * sin(psi);
+        R32 = cos(phi) * sin(theta) * sin(psi) - sin(phi) * cos(psi);
+        R33 = cos(phi) * cos(theta);
+
+    } else {
+        return -1;
+    }
 
-int rotmaty(float angle, Mat4 &R) {
-  R << cos(angle), 0.0f, -sin(angle), 0.0f, 0.0f, 1.0f, 0.0f, 0.0f,
-    sin(angle), 0.0f, cos(angle), 0.0f, 0.0f, 0.0f, 0.0f, 1.0f;
+    R << R11, R12, R13, R21, R22, R23, R31, R32, R33;
 
-  return 0;
+    return 0;
 }
 
-int rotmatz(float angle, Mat4 &R) {
-  R << cos(angle), -sin(angle), 0.0f, 0.0f, sin(angle), cos(angle), 0.0f,
-    0.0f, 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 1.0f;
+int quat2euler(Quaternion q, int euler_seq, Vec3 &euler) {
+    double qw, qx, qy, qz;
+    double qw2, qx2, qy2, qz2;
+    double phi, theta, psi;
+
+    qw = q.w();
+    qx = q.x();
+    qy = q.y();
+    qz = q.z();
+
+    qw2 = pow(qw, 2);
+    qx2 = pow(qx, 2);
+    qy2 = pow(qy, 2);
+    qz2 = pow(qz, 2);
+
+    if (euler_seq == 123) {
+        // euler 1-2-3
+        phi = atan2(2 * (qz * qw - qx * qy), (qw2 + qx2 - qy2 - qz2));
+        theta = asin(2 * (qx * qz + qy * qw));
+        psi = atan2(2 * (qx * qw - qy * qz), (qw2 - qx2 - qy2 + qz2));
+
+    } else if (euler_seq == 321) {
+        // euler 3-2-1
+        phi = atan2(2 * (qx * qw + qz * qy), (qw2 - qx2 - qy2 + qz2));
+        theta = asin(2 * (qy * qw - qx * qz));
+        psi = atan2(2 * (qx * qy + qz * qw), (qw2 + qx2 - qy2 - qz2));
+
+    } else {
+        return -1;
+    }
 
-  return 0;
+    euler << phi, theta, psi;
+    return 0;
 }
 
-void load_matrix(std::vector<double> x, int rows, int cols, MatX &y) {
-  int idx;
-
-  // setup
-  idx = 0;
-  y.resize(rows, cols);
+int quat2rot(Quaternion q, Mat3 &R) {
```

Comment:
(matrix()` is actually [synonymous](https://eigen.tuxfamily.org/dox/RotationBase_8h_source.html#l00045).)

OK, should we look for functions already covered by Eigen or STL (or Boost?) to avoid duplicating in this library?